### PR TITLE
Move rest end point generation and docker caching from HistomicsTK

### PR DESCRIPTION
**Pull request by cdeepakroy**
This PR moves the following functionality from HistomicsTK:
* Automatic REST end-point generation of slicer CLIs within Docker images
* Caching of Docker images and generation of REST end points for adding/deleting/updated docker images

### CODE REVIEW

File: plugin_tests/docker_test.py [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+
+from tests import base
+from girder import events
+
+# boiler plate to start and stop the server
+TIMEOUT = 180
+
+
+def setUpModule():
+    base.enabledPlugins.append('HistomicsTK')
```
Comment (cdeepakroy) [inline plugin_tests/docker_test.py]: Will this work? HistomicsTK is not a dependency of slicer_cli.

We have to check with @manthey / @zachmullen

Comment (cdeepakroy) [inline plugin_tests/docker_test.py]: See [here](https://github.com/girder/slicer_cli/blob/master/plugin.json#L5)

**Comment by cdeepakroy** [comment]
@salamb LGTM. We have to add travis-ci testing.

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    - source $build_path/miniconda/bin/activate $build_path/miniconda
+    - conda update --yes --all
+    - conda config --add channels https://conda.binstar.org/cdeepakroy
+
+install:
+    # https://github.com/pypa/pip/issues/2751
+    - conda install --yes libgfortran==1.0 setuptools==19.4 --file $main_path/requirements_c_conda.txt
+    - pip install -r $main_path/requirements.txt -r $main_path/requirements_c.txt
+    - cd $girder_path
+    - pip install -U -r requirements.txt -r requirements-dev.txt -r $main_path/requirements.txt -r $large_image_path/requirements.txt setuptools==19.4
```
Comment (cdeepakroy) [inline .travis.yml]: Lines 64-68 is all histomicstk specific stuff.

File: .travis.yml [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+compiler:
+    - gcc
+
+
+
+before_install:
+    
+    - GIRDER_VERSION=d910e905c4196e3046a129d0edb84e5c7e301596
+    - GIRDER_WORKER_VERSION=f834d4d3701df7f6f3f64fcd7d3f22d15b3b1db2
+    - LARGE_IMAGE_VERSION=60c32b11abb4d5f0d1696221a7e81ab21c611980
```
Comment (cdeepakroy) [inline .travis.yml]: We do not need large_image here.

File: .travis.yml (hunk 2 above)
Comment (cdeepakroy) [inline .travis.yml]: Just `pip install -r $main_path/requirements.txt` to install docker-py

File: server/__init__.py [hunk 4]
```diff
@@ -35,7 +35,7 @@ def load(info):
     resource = DockerResource('slicer_cli')
     info['apiRoot'].slicer_cli = resource
 
-    dockerImageModel = ModelImporter.model('dockerimagemodel', 'slicer_cli')
+    dockerImageModel = ModelImporter.model('docker_image_model', 'slicer_cli')
```
Comment (zachmullen) [inline server/__init__.py]: I'd recommend just using the name `docker_image`, by convention we do not need to include "model" in the name of models.

**Comment by cdeepakroy** [comment]
@manthey TravisCI is not scheduling the build for ever. Can you please take a look at the .travis.yml file?

**Comment by jbeezley** [comment]
I think because this is in the girder org, all of the PR's on girder/girder count against the priority in travis' queue.

**Comment by cdeepakroy** [comment]
@manthey Can you please take a look at .travis.yml if you can?

There are some [errors](https://travis-ci.org/girder/slicer_cli/builds/153586757) related to coverage.

**Pull request merged**